src/main/java/sc/fiji/pQCT/selectroi/RoiSelector.java

```diff
@@ -140,21 +151,44 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 					break;
 				}
 				// Check 4-connected neighbours
-				for (int j = y - 1; j <= y + 1; j+=2) {
-					for (int i = x - 1; i <= x + 1; i+=2) {
-						if (sieveTemp2[i + j * width] == 0) {
-							initialX.add(i);
-							initialY.add(j);
-						}
+				for (int nh = 0; nh<fourconnectedNHood.length;++nh) {
+					if (sieveTemp2[x+fourconnectedNHood[nh][0] + (y+fourconnectedNHood[nh][1])* width] == 0) {
+						initialX.add(x+fourconnectedNHood[nh][0]);
+						initialY.add(y+fourconnectedNHood[nh][1]);
 					}
+
 				}
+				
 			}
 			if (noLeak) {
-				sieveTemp = sieveTemp2;
+				sieveTemp = clone(sieveTemp2);
+				//Visualise sieve here
+				/*
+				ImagePlus tempImage2 = new ImagePlus("sieve");
+				tempImage2.setProcessor(new ByteProcessor(width,height,sieveTemp));
+				tempImage2.setDisplayRange(0,1);
+				tempImage2.show();
+				*/
 			}
 		}
 	}
 
+	protected double[] clone(double[] a){
+		double[] b = new double[a.length];
+		for (int i = 0;i<a.length;++i){
+			b[i] = a[i];
+		}
+		return b;
+	}
+	
+	protected byte[] clone(byte[] a){
+		byte[] b = new byte[a.length];
+		for (int i = 0;i<a.length;++i){
+			b[i] = a[i];
+		}
+		return b;
+	}
```

Comment:
I added a custom clone method because the .clone() did not actually produce a copy of the array but rather returned (in C-language terms) a pointer to the original array. When I then modified the contents of the copy, it also modified the original array I wanted not to modify. I think I've stumbled upon similar problems with Arrays.copyOf() in the past, and decided the only way I'll trust the code is by manually copying the contents to a new array. Having said that, System.arraycopy may establish the effect I wanted to have but I didn't remember that method when modfying the code.

Re protected, no particular reason, I've heard of the idea of trying to limit access if possible so used protected rather than public. I am not a programmer so I am inconsistent with these sorts of things...

Reply:
This is true for arrays of objects - they create what are called shallow copies. You get a new array, but the object references i.e. pointers point to the same place as the elements in the original array. However, with an array of primitives such as byte[] or double[] you do get a true copy. See: https://stackoverflow.com/questions/5821851/does-calling-clone-on-an-array-also-clone-its-contents

Reply:
`protected` is weaker than `private`, and it's actually quite rare that you really need it instead of private.